### ISSUE TITLE
No Module Found markupsafe

### ISSUE THREAD

**Issue by alifaff** [awslabs/chalice#78]
I am receiving the following error:  Unable to import module 'app': No module named markupsafe

Below is what I have in requirements.txt:
botocore==1.4.41
click==6.2
docutils==0.12
Flask==0.11.1
itsdangerous==0.24
MarkupSafe==0.23
Jinja2==2.8
jmespath==0.9.0
marshmallow==2.9.1
PyMySQL==0.7.6
python-dateutil==2.5.3
python-memcached==1.58
simplejson==3.8.2
six==1.10.0
SQLAlchemy==1.0.14
typing==3.5.1.0
Werkzeug==0.11.10

I am trying to get a simple example to work from a Flask app.  I have modified my code and it compiles without a problem, however when trying to access the service, I receive this error.

**Comment by jamesls** [comment]
Could you share steps to repro the issue?  I'm not able to repro.  Here's what I tried:

* Create a new project:

```
$ chalice new-project test-reqs
$ cd test-reqs
```

* I updated the `requirements.txt` file to match the exact contents you have above.

```
$ cat requirements.txt
botocore==1.4.41
click==6.2
docutils==0.12
Flask==0.11.1
itsdangerous==0.24
MarkupSafe==0.23
Jinja2==2.8
jmespath==0.9.0
marshmallow==2.9.1
PyMySQL==0.7.6
python-dateutil==2.5.3
python-memcached==1.58
simplejson==3.8.2
six==1.10.0
SQLAlchemy==1.0.14
typing==3.5.1.0
Werkzeug==0.11.10
```

*  Try to import `markupsafe` in my app:

```
$ cat app.py
from chalice import Chalice

import markupsafe

app = Chalice(app_name='test-reqs')


@app.route('/')
def index():
    return {'hello': 'world'}
```

* Deploy and test out the app:

```
$ chalice deploy
...
$ http https://....execute-api.us-west-2.amazonaws.com/dev/
HTTP/1.1 200 OK
Connection: keep-alive
Content-Length: 18
Content-Type: application/json
Date: Mon, 01 Aug 2016 20:05:08 GMT
x-amzn-RequestId: 3f19bd99-5823-11e6-8f93-8f6c60407675

{
    "hello": "world"
}

```

**Comment by alifaff** [comment]
Thanks for your prompt response. 

I have done the following - shorter version of the actual application, but even the bottom below does not seem to work - deploying to us-east-1:

virtualenv ~/.virtualenvs/reviewsreqs
source ~/.virtualenvs/reviewsreqs/bin/activate
pip install chalice

chalice new-project reviews-reqs
cd reviews-reqs/

pip install flask
pip install marshmallow
pip install simplejson
pip install sqlalchemy
pip install python-memcached

vi requirements.txt
```
botocore==1.4.41
click==6.2
docutils==0.12
Flask==0.11.1
itsdangerous==0.24
MarkupSafe==0.23
Jinja2==2.8
jmespath==0.9.0
marshmallow==2.9.1
PyMySQL==0.7.6
python-dateutil==2.5.3
python-memcached==1.58
simplejson==3.8.2
six==1.10.0
SQLAlchemy==1.0.14
typing==3.5.1.0
Werkzeug==0.11.10
```

vi app.py
```
import markupsafe
from chalice import Chalice

app = Chalice(app_name='reviews-reqs')

@app.route('/')
def index():
    return {'hello': 'world'}
```

Below is what CloudWatch has:

> START RequestId: 8d23b1fd-5830-11e6-9131-1f4f096f400a Version: $LATEST
> Unable to import module 'app': No module named markupsafe
> 
> END RequestId: 8d23b1fd-5830-11e6-9131-1f4f096f400a
> REPORT RequestId: 8d23b1fd-5830-11e6-9131-1f4f096f400a	Duration: 0.46 ms	Billed Duration: 100 ms 	Memory Size: 128 MB	Max Memory Used: 15 MB	

Anything I could be doing incorrectly?  I have tried to create from my mac as well as an EC2 instance, both seem to result in the same error.

**Comment by jamesls** [comment]
I followed the same steps you had above, and I'm still not able to get any errors.  I guess the next steps would be to see if we're actually creating the deployment package correctly.  Here's what I'd try next:

In your app directory, there should be a `.chalice` directory.  Within that directory there's a deployments directory that contains the zip file we created:

```
$ cd .chalice/deployments
$ ls
3d0d7f394c0cbf96ab351fc4d4b70491.zip
```

(Your filename hash might be different, that's ok.  There should still only be a single file in that directory).

I'd suggest trying to unzip the file, double check "markupsafe" is in that directory, and then try to import "app" from that directory:

```
$ mkdir tempdir
$ cp 3d0d7f394c0cbf96ab351fc4d4b70491.zip tempdir/
$ cd tempdir
$ unzip 3d0d7f394c0cbf96ab351fc4d4b70491.zip
# We should see markupsafe here.
$ ls | grep -i markupsafe
MarkupSafe-0.23.dist-info
markupsafe
```

And we should also be able to `import app` from this directory:

```
$ python
>>> import app
>>> app.app.routes
{'/': <chalice.RouteEntry object at 0x102e9bdd0>}
```

What do you get when you try the steps above?

**Comment by alifaff** [comment]
Thanks.  I see that three modules were not in the zip file: markupsafe, simplejson, and sqlalchemy.  Not sure why they are not in the zip file while yours are?

Below is what I get when I run app.app.routes:

> (clireviews)[ali@ip-172-31-70-156 tempdir]$ python
> Python 2.7.10 (default, Dec  8 2015, 18:25:23) 
> [GCC 4.8.3 20140911 (Red Hat 4.8.3-9)] on linux2
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import app
> >>> app.app.routes
> {'/': <chalice.RouteEntry object at 0x7f8609903c50>}

Is there something that I can do to force chalice to include these modules?

**Comment by jamesls** [comment]
In the above python repl (while you've cd'd into the zipfile directory), can you `import markupsafe`?

`chalice` is just running `pip install -r requirements.txt` (https://github.com/awslabs/chalice/blob/master/chalice/deployer.py#L699-L700), so I'm not sure why it would be skipping over those modules.  Perhaps pip thinks those modules are already installed somehow?  If the `import markupsafe` works in the python repl, perhaps it's being imported from somewhere else that's causing pip to already think it's installed.

**Comment by alifaff** [comment]
I was able to get it to work by doing the following (really odd why it worked):

1) Created the project
2) Updated the requirements.txt file
3) Updated app.py to import markupsafe
4) Tried to deploy the project as is and was warned that markupsafe was missing - note that I did not install any other dependencies from the requirements.txt file.  Since I could not deploy, I installed markupsafe via pip.  I did not install anything else and then tried to redeploy the chalice project and it worked.

Really odd as to why markupsafe does not install by simply reading the requirements.txt.  

Thanks for your help!
